### PR TITLE
Enumerable Module

### PR DESCRIPTION
**Pull request by HillaryOkello**
In this project, we created the following ruby enumerable methods. 
* my_each
* my_each_with_index
* my_select
* my_all?
* my_any?
* my_none?
* my_count
* my_map
* my_inject
* multiply_els

### CODE REVIEW

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  end
+
+  # 3.my_select
+  def my_select
+    select_arr_result = []
+    each do |item|
+      select_arr_result << item if yield(item)
+    end
+    select_arr_result
+  end
```
Comment (PhillipUg) [inline main.rb]: - [ ] These three methods all should return an `Enumerator` when no block is given

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+  # 8.my_map
+  def my_map(parameter = nil)
+    # puts parameter
+    result_arr = []
+    my_each {|x| result_arr << yield(x)} if parameter.nil?
+    
+    my_each {|x| result_arr << parameter.call(x)} unless parameter.nil?
+    result_arr
+  end
```
Comment (PhillipUg) [inline main.rb]: - [ ] Should return an `Enumerator` when no block is given

**Review by PhillipUg — CHANGES_REQUESTED** [review]
Hey team, 👋

You have done an amazing job implementing the project requirements of this project. Your `my_count`, `my_inject`, and `my_none` methods work very well.

However, there are still some changes 👇 you need to implement to approve this request 🔍

### Changes Required ♻️ 

- [ ] You have some Linter errors. Please always ensure you have no linter errors before submitting for review in the future.

- [x] Your `my_all?`, `my_any?`, and `multiply_els` need more work. So I am going to provide you with some custom test scenarios. Please make sure your methods pass all of them before submitting for re-review.

```

# 4. my_all? (example test cases)
puts 'my_all?'
puts '-------'
p [3, 5, 7, 11].my_all?(&:odd?) # => true
p [-8, -9, -6].my_all? { |n| n < 0 } # => true
p [3, 5, 8, 11].my_all?(&:odd?) # => false
p [-8, -9, -6, 0].my_all? { |n| n < 0 } # => false
# test cases required by tse reviewer
p [1, 2, 3, 4, 5].my_all? # => true
p [1, 2, 3].my_all?(Integer) # => true
p %w[dog door rod blade].my_all?(/d/) # => true
p [1, 1, 1].my_all?(1) # => true
puts

# 5. my_any? (example test cases)
puts 'my_any?'
puts '-------'
p [7, 10, 3, 5].my_any?(&:even?) # => true
p [7, 10, 4, 5].my_any?(&:even?) # => true
p %w[q r s i].my_any? { |char| 'aeiou'.include?(char) } # => true
p [7, 11, 3, 5].my_any?(&:even?) # => false
p %w[q r s t].my_any? { |char| 'aeiou'.include?(char) } # => false
# test cases required by tse reviewer
p [3, 5, 4, 11].my_any? # => true
p [1, nil, false].my_any?(1) # => true
p [1, nil, false].my_any?(Integer) # => true
p %w[dog door rod blade].my_any?(/z/) # => false
p [1, 2, 3].my_any?(1) # => true
puts

# puts multiply_els([2, 4, 5]) # => 40
# puts
```

#### Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack as **Phillip Musiime**
I would be happy to help! 👩‍💻 

Happy coding 🤓

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        true_counter += 1 if yield(x)
+      end
+      true_counter == length_array
+    else
+      my_each do |element|
+        true_counter1 += 1 if !element.nil? == true
+      end
+      true_counter1 == length_array
+    end
+  end
```
Comment (elmejdki) [inline main.rb]: ## my_all?:

- [ ] when no block or argument is given returns false when one of the collection members is false or nil:

```ruby

      false_array = [1, false, 'hi', []]
      false_array.my_all? === false_array.all? => false

```

- [ ] when a class is passed as an argument returns false if any of the collection is not a member of such class:

```ruby

      array = ["word", 23, 3, 98]
      array.my_all? === array.all? => false

```

- [ ] when a Regex is passed as an argument returns false if any of the collection does not match the Regex:

```ruby

      array = ["word1", "word2", "word3"]
      words.my_all?(/o/) === words.all?(/o/) => false

```

- [ ] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern:

```ruby

      array = [21, 22, 54, 34]
      array.my_all?(3) === array.all?(3) => false

```

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      when Regexp
+        my_each { |x| return true if x =~ parameter }
+      else
+        my_each do |element|
+          initial_value = true unless element.nil?
+        end
+      end
+      initial_value
+    end
+  end
```
Comment (elmejdki) [inline main.rb]: ## my_any?

_use the same variables in the previous examples_

- [ ] when no block or argument is given returns false if at least one of the collection is not true:

```ruby

    false_array.my_any? === false_array.any? => false

```

- [ ] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer:

```ruby

    words.my_any?(Integer) === words.any?(Integer) => false

```

- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern:

```ruby

    words.my_any?('cat') === words.any?('cat') => false

```

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 
## Great job so far @team, you are almost there we still have some small improvements, and you are done.

- [ ] Please fix the linter errors, there are some errors related to RSpec, if you can't fix it please let me know I will be happy to help.

I have left some comments below please check them.

### Please make all the required changes before you request another code review.
## Happy Coding :tophat: 
_If you have any question or you feel stuck don't hesitate to contact me on slack_

**Review by PhillipUg — APPROVED** [review]
# Status: Project Approved! 🥇 
Congratulations on implementing all project requirements and getting this project approved. :+1:

You can merge this branch into the development branch when you're ready and go on to your next project.
:bulb:  Do not forget to submit your project completion form after merging this branch.

All the best team,

For any assistance, you can contact me on slack as **Phillip Musiime**

**Pull request merged**